src/Reflex/Dom/Xhr.hs

```diff
@@ -163,20 +180,31 @@ newXMLHttpRequestWithError req cb = do
       statusText <- liftIO $ xmlHttpRequestGetStatusText xhr
       when (readyState == 4) $ do
         t <- if rt == Just XhrResponseType_Text || isNothing rt
-               then liftIO $ xmlHttpRequestGetResponseText xhr
-               else return Nothing
+             then liftIO $ xmlHttpRequestGetResponseText xhr
+             else  return Nothing
         r <- liftIO $ xmlHttpRequestGetResponse xhr
-        _ <- liftIO $ cb $ Right $ XhrResponse
-          { _xhrResponse_status = status
-          , _xhrResponse_statusText = statusText
-          , _xhrResponse_response = r
-          , _xhrResponse_responseText = t
-          }
+        h <- case _xhrRequestConfig_responseHeaders c of
+          AllHeaders -> liftIO $ parseAllHeadersString <$>
+            xmlHttpRequestGetAllResponseHeaders xhr
+          OnlyHeaders xs -> liftIO $ Map.fromList <$> traverse
+            (\x -> (x,) <$> xmlHttpRequestGetResponseHeader xhr x) xs
```

Comment:
If you're using a Set for xs, you can use [Map.fromSet](https://hackage.haskell.org/package/containers-0.5.7.1/docs/Data-Map-Lazy.html#v:fromSet) to avoid needing to sort the keys again.  I'd recommend just doing something like Map.fromSet id and then traversing the result.